In [19]:
import numpy as np
import pandas as pd
from pandas import Series, DataFrame

import sklearn
from sklearn import preprocessing
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn import metrics 
from sklearn.metrics import classification_report
from sklearn.metrics import mean_squared_error
from pylab import rcParams

import matplotlib.pyplot as plt
import seaborn as sb

## Multinomial logistic regression and cross validation

In [3]:
drink = pd.read_csv('strongdrink.txt')
drink.sample(5)

cultivar   alco  malic   ash   alk  magn  tot_phen  flav  nonfl_phen  \
59          2  12.37   0.94  1.36  10.6    88      1.98  0.57        0.28   
151         3  13.11   1.90  2.75  25.5   116      2.20  1.28        0.26   
38          1  13.07   1.50  2.10  15.5    98      2.40  2.64        0.28   
29          1  14.02   1.68  2.21  16.0    96      2.65  2.33        0.26   
7           1  14.06   2.15  2.61  17.6   121      2.60  2.51        0.31   

     proanth  color_int   hue  OD280rat  proline  
59      0.42       1.95  1.05      1.82      520  
151     1.56       7.10  0.61      1.33      425  
38      1.37       3.70  1.18      2.69     1020  
29      1.98       4.70  1.04      3.59     1035  
7       1.25       5.05  1.06      3.58     1295

In [4]:
X = drink[['alco', 'malic', 'tot_phen', 'color_int']].values
y = drink['cultivar'].values

In [14]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25,random_state=20)
clf = LogisticRegression(random_state=0, solver='lbfgs',
                         multi_class='multinomial').fit(X_train, y_train)

parameters = pd.DataFrame({"j=1":np.append(clf.intercept_[0],clf.coef_[0]),
                           "j=2":np.append(clf.intercept_[1],clf.coef_[1])},
                          index=["β0","β1",'β2','β3','β4'])

parameters

j=1        j=2
β0 -24.027617  22.780733
β1   1.701734  -1.466297
β2  -0.265788  -0.332951
β3   1.224101   0.663556
β4   0.022507  -0.922682

In [17]:
y_pred=clf.predict(X_test)
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           1       0.87      1.00      0.93        13
           2       1.00      0.90      0.95        21
           3       1.00      1.00      1.00        10

    accuracy                           0.95        44
   macro avg       0.96      0.97      0.96        44
weighted avg       0.96      0.95      0.96        44



In [18]:
drink['cultivar'].value_counts()

2    71
1    59
3    46
Name: cultivar, dtype: int64

In [21]:
print(mean_squared_error(y_test, y_pred))

0.045454545454545456


- Error rates for j=1 is 13%, j=2 is 0%.
- category 2 is best predicted as category 3 is the baseline class.
- The most accurately predicdted category is not the one with the most observations.
- MSE is 0.04545.

### No time to finish the Problemset in this week, will finish it when I am back to Australia.